keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,657 @@
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: admin
+roleRef:
+  kind: ClusterRole
+  name: admin
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: namespace-owner
+roleRef:
+  kind: Role
+  name: namespace-owner
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Issue `#2` can be solved using a `MutatingAdmissionWebhook` which adds an annotation
+
+```
+authorization.k8s.io/creator=name
+```
+
+during the API request to create a namespace.
+Then the controller is able to fill `$(CREATOR)` from that annoation.
+
+This annotation must be made immutable.
+The `MutatingAdmissionWebhook` can easily achieve this in further _update/patch_ requests.
+Alternatively, a `ValidatingAdmissionWebhook` can be used to reject requests which attempts to change the annotation,
+however it may break existing client implementations which don't expect immutable annotations as they don't exist in Kubernetes today.
+
+Issue `#3` can be solved by using a `MutatingAdmissionWebhook` to adding `NamespaceTemplate` enforcement controller into `metadata.initializers` during the creation of the namespace.
+By default, the namespace creator will hold the response for a period of time until the namespace is fully initialized,
+(ref: `WaitForInitialized` in 
+[k8s.io/apiserver/pkg/registry/generic/registry/store.go](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go))
+unless `includeUninitialized` is set to `true` in the request from the client.
+
+For further protection,
+to prevent other privileged users/service accounts to create objects into an uninitialized namespace,
+A `ValidatingAdmissionWebhook` can be used to validate with authorization interface if the requestor
+has the `initialize` permission on the current namespace if it's uninitialized:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-initialization
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    verbs: ["initialize"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: namespace-initializer
+roleRef:
+  kind: ClusterRole
+  name: namespace-initialization
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: ServiceAccount
+  name: "system:serviceaccount:kube-system:name"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+#### Extending Initializers
+
+Kubernetes provides resource initialization mechanism
+(injecting items into `metadata/initializers.pending` list) in a _MutatingAdmissionWebhook_ to
+allow controllers to finish initialization work before the creation request is completed and response is 
+delivered to the client.
```

Comment:
Last I heard initializers were alpha not going to move forward.  Am I reading right that this is depending on those?

Reply:
This doesn't depends on Initializers admission controller (alpha feature). It's proposed to use the metadata.initializers mechanism: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

Reply:
It isn't clear to me (and maybe @smarterclayton can shine a light) that the admission controller and the `metadata.initializers` mechanism are separate.  My assumption/understanding is that these were both alpha.  Is the mechanism you are relying on really GA?